XenModel/Actions/VM/VMSnapshotDeleteAction.cs

```diff
@@ -29,33 +29,35 @@
  * SUCH DAMAGE.
  */
 
-using System;
 using XenAdmin.Actions.VMActions;
 using XenAPI;
 
 
 namespace XenAdmin.Actions
 {
-    public class VMSnapshotDeleteAction : PureAsyncAction
+    public class VMSnapshotDeleteAction : AsyncAction
     {
 
         private VM m_Snapshot;
         public VMSnapshotDeleteAction(VM snapshot)
-            : base(snapshot.Connection, String.Format(Messages.ACTION_VM_DELETE_SNAPSHOT_TITLE, snapshot.Name()))
+            : base(snapshot.Connection, string.Format(Messages.ACTION_VM_DELETE_SNAPSHOT_TITLE, snapshot.Name()))
         {
-            this.VM = Connection.Resolve<VM>(snapshot.snapshot_of);
-            this.m_Snapshot = snapshot;
+            VM = Connection.Resolve(snapshot.snapshot_of);
+            m_Snapshot = snapshot;
+
+            if (m_Snapshot.power_state == vm_power_state.Suspended)
+                ApiMethodsToRoleCheck.Add("VM.hard_shutdown");
         }
 
         protected override void Run()
         {
-            Description = String.Format(Messages.ACTION_VM_DELETE_SNAPSHOT_TITLE, m_Snapshot.Name());
+            Description = string.Format(Messages.ACTION_VM_DELETE_SNAPSHOT_TITLE, m_Snapshot.Name());
             if (m_Snapshot.power_state == vm_power_state.Suspended)
             {
-                XenAPI.VM.hard_shutdown(Session, m_Snapshot.opaque_ref);
+                VM.hard_shutdown(Session, m_Snapshot.opaque_ref);
             }
             VMDestroyAction.DestroyVM(Session, m_Snapshot, true);

```

Comment:
The `VMSnapshotDeleteAction` can be corrected either by listing the additional `ApiMethodsToRoleCheck` or by deriving from `DestroyVMAction`; I prefer doing the latter.

`CreateVMAction` is tricky because the `VMDestroyAction` is only called in case of an error, so it's not great to check the additional API calls without knowing if they will be needed at all. Note that we don't seem to care too much if the clean up operation fails, so I suggest we skip the RBAC check. In most cases we should be ok, as normally the creation and destruction operations in xapi need similar privileges.

Reply:
Since it's a cleanup action, I'm also ok with skiping 👍